flow/connectors/clickhouse/normalize.go

```diff
@@ -265,46 +267,82 @@ func (c *ClickHouseConnector) generateCreateTableSQLForNormalizedTable(
 	return result, nil
 }
 
-// Returns a list of order by columns ordered by their ordering, and puts the pkeys at the end.
-// pkeys are excluded from the order by columns.
+// getOrderedOrderByColumns returns columns to be used for ordering in destination table operations.
+// If no custom ordering is specified, returns the source table's primary key columns as ordering keys.
+// If custom ordering columns are specified, returns those columns sorted by their ordering value.
+// The boolean return value indicates whether ordering keys contain nullable keys.
 func getOrderedOrderByColumns(
 	tableMapping *protos.TableMapping,
+	sourceColumns []*protos.FieldDescription,
 	sourcePkeys []string,
 	colNameMap map[string]string,
-) []string {
-	pkeys := make([]string, len(sourcePkeys))
-	for idx, pk := range sourcePkeys {
-		pkeys[idx] = peerdb_clickhouse.QuoteIdentifier(pk)
-	}
-	if len(sourcePkeys) > 0 && len(colNameMap) > 0 {
-		for idx, pk := range sourcePkeys {
-			pkeys[idx] = peerdb_clickhouse.QuoteIdentifier(getColName(colNameMap, pk))
-		}
+	tableNullableEnabled bool,
+) ([]string, bool) {
+	columnNullableMap := make(map[string]bool)
+	for _, col := range sourceColumns {
+		columnNullableMap[col.Name] = col.Nullable
 	}
 
-	orderby := make([]*protos.ColumnSetting, 0)
+	columnNullableEnabledMap := make(map[string]bool)
+	columnOrderingMap := make(map[string]int32)
 	if tableMapping != nil {
 		for _, col := range tableMapping.Columns {
+			columnNullableEnabledMap[col.SourceName] = col.NullableEnabled
 			if col.Ordering > 0 {
-				orderby = append(orderby, col)
+				columnOrderingMap[col.SourceName] = col.Ordering
 			}
 		}
 	}
 
-	if len(orderby) == 0 {
-		return pkeys
+	isNullableAndNullableEnabled := func(columnName string) bool {
+		nullable, exists := columnNullableMap[columnName]
+		if !exists {
+			return false
+		}
+		nullableEnabled := tableNullableEnabled || columnNullableEnabledMap[columnName]
+
+		return nullable && nullableEnabled
+	}
+
+	hasNullableKeys := false
+
+	// Case 1: No custom ordering - use primary keys
+	if len(columnOrderingMap) == 0 {
+		pkeys := make([]string, len(sourcePkeys))
+		for idx, pk := range sourcePkeys {
+			pkeys[idx] = peerdb_clickhouse.QuoteIdentifier(pk)
+			if !hasNullableKeys && isNullableAndNullableEnabled(pk) {
```

Comment:
isNullableAndNullableEnabled should always return false for mysql/pg/mongo as source, but checking this for completion in case we add sources that contain nullable columns